python/tskit/trees.py

```diff
@@ -5216,30 +5295,33 @@ def alignments(self, *, reference_sequence=None, missing_data_character=None):
         """
         if not self.discrete_genome:
             raise ValueError("sequence alignments only defined for discrete genomes")
-
+        left, right = self._check_genomic_range(left, right, ensure_integer=True)
         missing_data_character = (
             "N" if missing_data_character is None else missing_data_character
         )
 
-        L = int(self.sequence_length)
+        L = int(right - left)
         a = np.empty(L, dtype=np.int8)
         if reference_sequence is None:
             if self.has_reference_sequence():
                 # This may be inefficient - see #1989. However, since we're
                 # n copies of the reference sequence anyway, this is a relatively
-                # minor tweak.
-                reference_sequence = self.reference_sequence.data
+                # minor tweak. We may also want to recode the below not to use direct
+                # access to the .data attribute, e.g. if we allow reference sequences
+                # to start at non-zero positions
+                reference_sequence = self.reference_sequence.data[left:right]
```

Comment:
Ah, I was worried about negative indexing here, but that's checked in `_check_genomic_range`.